cpg-core/src/main/java/de/fraunhofer/aisec/cpg/frontends/LanguageTraits.kt

```diff
@@ -43,3 +44,68 @@ interface HasTemplates : LanguageTrait
  * functions.
  */
 interface HasDefaultArguments : LanguageTrait
+
+/**
+ * A language trait that specifies that this language has a complex call resolution that we need to
+ * fine-tune in the language implementation.
+ */
+interface HasComplexCallResolution : LanguageTrait {
+    // TODO: better name?
+    fun doBetterCallResolution()
+}
+
+/** A language trait that specifies if the language supports function pointers. */
+interface HasFunctionPointers : LanguageTrait
+
+/**
+ * A language trait that specifies if the language has the concept of "structs". The alternative is
+ * to use classes. Note that some languages can have both and that in some languages structs and
+ * classes are essentially the same. This is mostly used to determine if an inferred record
+ * declaration *can be* a struct or *must be* a class.
+ */
+interface HasStructs : LanguageTrait
+
+/**
+ * A language trait that specifies if the language has the concept of "classes". The alternative is
+ * to use structs. Note that some languages can have both and that in some languages structs and
+ * classes are essentially the same. This is mostly used to determine if an inferred record
+ * declaration *can be* a class or *must be* a struct.
+ */
+interface HasClasses : LanguageTrait
+
+/**
+ * A language trait, that specifies that this language has support for superclasses. If so, we
+ * should consider the specified superclass keyword to resolve calls etc.
+ */
+interface HasSuperclasses : LanguageTrait {
+    /**
+     * Determines which keyword is used to access functions, etc. of the superclass of an object
+     * (often "super).
+     */
+    val superclassKeyword: String
+}
+
+/**
+ * A language trait, that specifies that this language has certain qualifiers. If so, we should
+ * consider them when parsing the types.
+ */
+interface HasQualifier : LanguageTrait {
+    /** The qualifiers which exist in the language. */
+    val qualifiers: List<String>
+}
+
+/**
+ * A language trait, that specifies that this language has "elaborated type specifiers". If so, we
+ * should consider them when parsing the types.
+ */
+interface HasElaboratedTypeSpecifier : LanguageTrait {
+    val elaboratedTypeSpecifier: List<String>
+}
+
+/**
+ * A language trait, that specifies that this language has specifiers which let us conclude that we
+ * do not know the type. If so, we should consider them when parsing the types.
+ */
+interface HasUnknownType : LanguageTrait {
```

Comment:
Do you mean something like `auto` or `var`? This is different from an unknown type I would say.

Reply:
Yes, that's what I mean. I'm open for better names

Reply:
OK I guess its not 100 % the same as a truly unknown type because we cannot see its declaration, etc. but we will currently treat it as such for legacy reasons.